### ISSUE TITLE
Custom AssetSource(s) must be registered before DefaultPlugins in order to work

### ISSUE THREAD

**Issue by 66OJ66** [bevyengine/bevy#10162]
## Bevy version

0.12-dev

## What you did

I attempted to register an AssetSource:

```rust
fn main() {
    App::new()
        .add_plugins(
            DefaultPlugins.set(AssetPlugin {
                mode: AssetMode::ProcessedDev,
                ..default()
            }),
        )
        .register_asset_source(
            "test",
            AssetSource::build()
                .with_reader(|| Box::new(FileAssetReader::new("test")))
        )
        .run();
}
```

## What went wrong

When I tried to load a handle from that AssetSource, I got this error:
```rust
ERROR bevy_asset::server: Asset Source 'AssetSourceId::Name(test)' does not exist
```

## Additional information

Registering the asset source _before_ adding the `DefaultPlugins` resolved the issue - this seems a bit counter-intuitive though.

I think there's broadly 2 options to stop others encountering this issue:

1) Mention this ordering requirement in the `register_asset_source` comments:
https://github.com/bevyengine/bevy/blob/6b070b1776c63aa69cc60185e04c293b32e60164/crates/bevy_asset/src/lib.rs#L244-L245

2) Call `build_sources` from `Plugin::finish()` instead of `Plugin::build()`
https://github.com/bevyengine/bevy/blob/6b070b1776c63aa69cc60185e04c293b32e60164/crates/bevy_asset/src/lib.rs#L139-L140
This might not work for anyone attempting to load handles from a custom `AssetSource` in a `FromWorld` impl, although a similar issue was resolved in https://github.com/bevyengine/bevy/pull/9429

**Comment by alice-i-cecile** [comment]
This is a special case of https://github.com/bevyengine/bevy/issues/1255.

**Comment by andriyDev** [comment]
Some relevant discussion in Discord: https://discordapp.com/channels/691052431525675048/1384599676006826097/1384642518871838740

TL;DR there are cases where you may want to change the set of asset sources at runtime (e.g., loading assets from an "archive" #12279, which the user selects at runtime).

**Comment by kristoff3r** [comment]
> Some relevant discussion in Discord: https://discordapp.com/channels/691052431525675048/1384599676006826097/1384642518871838740
> 
> TL;DR there are cases where you may want to change the set of asset sources at runtime (e.g., loading assets from an "archive" [#12279](https://github.com/bevyengine/bevy/issues/12279), which the user selects at runtime).

I spent 30 minutes looking at what it would take to achieve this, and it's a bit invasive unfortunately. The only option I can see is to wrap `AssetSources` inside `AssetServerData` in an `RwLock`, put the individual `AssetSource`s inside `Arc` and then change all the APIs for asset sources to use `Arc` instead of references. That's a bit too invasive for me to PR, so I'll just work around this limitation for now.

**Comment by andriyDev** [comment]
@kristoff3r I was thinking about this a little more, and there is a workaround here where you have a custom AssetSource holding `Arc<Mutex<Option<InternalState>>>`. Then at runtime you can mess with that internal state. It still means you're stuck to a fixed number of asset sources, but with some cleverness you can make it work (e.g., "my_runtime_source://the_fake_source/my/asset/path").

**Comment by kristoff3r** [comment]
> [@kristoff3r](https://github.com/kristoff3r) I was thinking about this a little more, and there is a workaround here where you have a custom AssetSource holding `Arc<Mutex<Option<InternalState>>>`. Then at runtime you can mess with that internal state. It still means you're stuck to a fixed number of asset sources, but with some cleverness you can make it work (e.g., "my_runtime_source://the_fake_source/my/asset/path").

That wouldn't allow me to borrow from the internal state in the reader I return I think? The library I use to read the archive currently requires that. I could probably hack around it, but in that case I'd rather just load on startup like you suggested instead.
